### PR TITLE
GG-21926: Export node configuration to gmc.

### PR DESCRIPTION
**Pull request by smoldachev**

### CODE REVIEW

File: modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationService.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    /** Manager. */
+    private final WebSocketManager mgr;
+
+    /** Sender. */
+    private final GmcSender<VisorGridConfiguration> snd;
+
+    /** On node traces listener. */
+    private final IgniteBiPredicate<UUID, Object> lsnr = this::onNodeConfiguration;
+
+    public NodeConfigurationService(GridKernalContext ctx, WebSocketManager mgr) {
```
Comment (akuznetsov-gridgain) [inline modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationService.java]: Missing javadocs.

File: modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationService.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    @Override public void close() {
+        ctx.grid().message().stopLocalListen(NODE_CONFIGURATION_TOPIC, lsnr);
+        U.closeQuiet(snd);
+    }
+
+    /**
+     * @param uuid Uuid.
+     * @param cfgList Config list.
+     */
+    boolean onNodeConfiguration(UUID uuid, Object cfgList) {
```
Comment (akuznetsov-gridgain) [inline modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationService.java]: maybe 'boolean' -> 'Void'  + return null ?

File: modules/gmc-agent/src/main/java/org/gridgain/service/sender/RetryableSender.java [hunk 3]
```diff
@@ -22,6 +22,7 @@
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+import com.google.common.collect.Lists;
```
Comment (akuznetsov-gridgain) [inline modules/gmc-agent/src/main/java/org/gridgain/service/sender/RetryableSender.java]: Do we really need guava ?

File: modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationExporter.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    /** Status description. */
+    public static final String NODE_CONFIGURATION_TOPIC = "gmc-node-configuration-topic";
+
+    /** Sender. */
+    private final CoordinatorSender<VisorGridConfiguration> snd;
+
+    /** Context. */
+    private GridKernalContext ctx;
+
```
Comment (nva) [inline modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationExporter.java]: not needed line

File: modules/gmc-agent/src/main/java/org/gridgain/service/sender/RetryableSender.java [hunk 5]
```diff
@@ -73,6 +74,14 @@ protected RetryableSender(int cap) {
         exSrvc.shutdown();
     }
 
+    /**
+     * @param element Element to send.
+     */
+    public void send(T element) {
+        if (element != null)
+            addToQueue(Lists.newArrayList(element));
```
Comment (ingvard) [inline modules/gmc-agent/src/main/java/org/gridgain/service/sender/RetryableSender.java]: Collections.singleton()

File: modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationService.java (hunk 1 above)
Comment (smoldachev) [reply modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationService.java]: Fix.

File: modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationService.java (hunk 2 above)
Comment (smoldachev) [reply modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationService.java]: The "onNodeConfiguration" is IgniteBiPredicate with "boolean apply" method.

File: modules/gmc-agent/src/main/java/org/gridgain/service/sender/RetryableSender.java (hunk 3 above)
Comment (smoldachev) [reply modules/gmc-agent/src/main/java/org/gridgain/service/sender/RetryableSender.java]: Nope, change on Collections.singletonList.

File: modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationExporter.java (hunk 4 above)
Comment (smoldachev) [reply modules/gmc-agent/src/main/java/org/gridgain/service/config/NodeConfigurationExporter.java]: Fix.

File: modules/gmc-agent/src/main/java/org/gridgain/service/sender/RetryableSender.java (hunk 5 above)
Comment (smoldachev) [reply modules/gmc-agent/src/main/java/org/gridgain/service/sender/RetryableSender.java]: Thanks :)

File: modules/gmc-agent/src/main/java/org/gridgain/agent/Agent.java [hunk 6]
```diff
@@ -103,12 +111,14 @@ public Agent(GridKernalContext ctx) {
...
     @Override public void onKernalStart(boolean active) {
         spanExporter = new GmcSpanExporter(ctx);
+        nodeConfigurationExporter = new NodeConfigurationExporter(ctx);
         metaStorage = ctx.cache().context().database().metaStorage();
 
         launchAgentListener(null, ctx.discovery().discoCache());
 
         // Listener for coordinator changed.
         ctx.event().addDiscoveryEventListener(this::launchAgentListener, EVTS_DISCOVERY);
+        nodeConfigurationExporter.export();
```
Comment (akuznetsov-gridgain) [inline modules/gmc-agent/src/main/java/org/gridgain/agent/Agent.java]: I think it is not good idea to do potentially "heavy" operations in onKernalStart method. 
May be backend should "poll" for configurations from cluster?

Comment (smoldachev) [reply modules/gmc-agent/src/main/java/org/gridgain/agent/Agent.java]: I don't think what putting element in queue is "heavy" operation (yes, we are wrapping configuration but in the future we will have a serializable IgniteConfiguration). Maybe we can suggest an additional type of callbacks (afterStart for example) to core team?

**Pull request merged**